src/libcrun/criu.c

```diff
@@ -218,6 +218,88 @@ libcrun_container_checkpoint_linux_criu (libcrun_container_status_t *status,
   return 0;
 }
 
+static int
+prepare_restore_mounts (runtime_spec_schema_config_schema *def,
+                        char *root,
+                        libcrun_error_t *err)
+{
+  int i;
+
+  /* Go through all mountpoints to be able to recreate missing mountpoints. */
+  for (i = 0; i < def->mounts_len; i++)
+    {
+      char *dest = def->mounts[i]->destination;
+      char *type = def->mounts[i]->type;
+      cleanup_close int root_fd = -1;
+      bool on_tmpfs = false;
+      int is_dir = 1;
+      size_t j;
+
+      /* cgroup restore should be handled by CRIU itself */
+      if (strcmp (type, "cgroup") == 0
+          || strcmp (type, "cgroup2") == 0)
+        continue;
+
+      for (j = 0; j < def->mounts_len; j++)
+        {
+          cleanup_free char *dest_loop = NULL;
+
+          xasprintf (&dest_loop, "%s/", def->mounts[j]->destination);
+          /* Check if the parent mount is on a tmpfs. CRIU restores
+           * all tmpfs. We do need to recreate directories on a tmpfs. */
+          if (strncmp (dest, dest_loop, strlen (dest_loop)) == 0 &&
+              strcmp (def->mounts[j]->type, "tmpfs") == 0)
+            {
+              /* This is a mountpoint which is below a tmpfs.*/
+              on_tmpfs = true;
+              continue;
+            }
+        }
+
+      if (on_tmpfs)
+        continue;
+
+      /* For bind mounts check if the source is a file or a directory. */
+      for (j = 0; j < def->mounts[i]->options_len; j++)
+        {
+          if (strcmp (def->mounts[i]->options[j], "bind") == 0
+              || strcmp (def->mounts[i]->options[j], "rbind") == 0)
+            {
+              is_dir = crun_dir_p (def->mounts[i]->source, false, err);
+              if (UNLIKELY (is_dir < 0))
+                return is_dir;
+            }
+        }
+
+      root_fd = open (root, O_RDONLY | O_CLOEXEC);
+      if (UNLIKELY (root_fd == -1))
+        return crun_make_error (err, errno,
+                                "error opening container root directory %s\n",
+                                root);
+
+      if (is_dir)
+        {
+          int ret;
+
+          ret = crun_safe_ensure_directory_at (root_fd, root, strlen (root),
+                                               dest, 0755, err);
+          if (UNLIKELY (ret < 0))
+            return ret;
+        }
+      else
+        {
+          int ret;
+
+          ret = crun_safe_ensure_file_at (root_fd, root, strlen (root), dest,
```

Comment:
Frankly I don't remember runc creating files, only directories.

The only way such files might appear here I can think of are masked path under `/proc` or `/sys`, and obviously those do not need to be created.

Am I missing something?

Reply:
Yes, runc does not create files, that is correct, but as crun at other places in the code did differentiate between files and directories I also used it here. One example is `/run/.containerenv`. I guess it is not really necessary, but I was just following existing crun code.